harmony_py/auth.py

```diff
@@ -0,0 +1,197 @@
+    """
+
+    if username and password:
+        return SessionWithHeaderRedirection(username, password)
+    elif username and not password:
+        password = getpass()
+        return SessionWithHeaderRedirection(username, password)
+    elif netrc_file:
+        return SessionWithHeaderRedirection()
+    else:
+        if cfg.EDL_USERNAME and cfg.EDL_PASSWORD:
+            return SessionWithHeaderRedirection(cfg.EDL_USERNAME, cfg.EDL_PASSWORD)
+        else:
+            raise MissingCredentials('Authentication: No credentials found.')
+
+
+def authenticate(username: Optional[str] = None, password: Optional[str] = None,
+                 netrc_file: Optional[bool] = False,
+                 verify: Optional[bool] = True) -> Session:
+    """
+    Create are requests-futures session for authenticated HTTP calls after optionally verifying
+    credentials.
+    Attempts to create an authenticated session in the following order:
+    1) If ``username`` and ``password`` are not None, create a session.
+    2) If ``username`` is specified but not ``password``, prompt user for a password and create a
+    session.
+    3) If ``netrc_file`` is True, rely on the automatic behavior of requests to find relevant
+    credentials in a .netrc file.
+    4) Attempt to read a username and password from environment variables, either from the system
+    or from a .env file to return a session.
+
+    Parameters:
+        username (str, optional): The EDL username.
+        password (str, optional): The EDL password.
+        netrc_file (bool, optional): Whether a .netrc file should be preferred for credentials.
+        verify (bool, optional): Whether EDL credentials will be verified.
+
+    Returns:
+        (:obj:`SessionWithHeaderRedirection`): The authenticated requests session.
+
+    :raises MissingCredentials: No credentials were specified or found.
+    :raises BadAuthentication: Incorrect credentials or unknown error.
+    """
+    edl_verification_url = cfg.EDL_VERIFICATION_URL
+    num_workers = int(cfg.NUM_REQUESTS_WORKERS)
+
+    session = _authenticate(username=username, password=password, netrc_file=netrc_file)
+    futures_session = FuturesSession(session=session,
+                                     executor=ThreadPoolExecutor(max_workers=num_workers))
+
+    if not verify:
+        return futures_session
+    else:
+        result = (futures_session.get(edl_verification_url)).result()
+        if result.status_code == 200:
+            return futures_session
+        elif result.status_code == 401:
+            raise BadAuthentication('Authentication: incorrect or missing credentials during '
+                                    'credential verification.')
+        else:
+            raise BadAuthentication('Authentication: An unknown error occurred during credential '
+                                    f'verification: HTTP {result.status_code}')
+
+
+# if __name__ == "__main__":
```

Comment:
I'd remove commented out code. If this is meant as an example should we put it in a Jupyter notebook or some other Python mechanism for examples?

Reply:
I'm sorry - that's my bad. I had meant to remove that before pushing. A python notebook was included for example purposes, too.